### DISCUSSION TITLE
RTU and PLC's network do not connect?

### DISCUSSION THREAD

**Discussion by samuelh2005** [MikaylaFischler/cc-mek-scada#376] [Support] [answered]
# Introduction

So I've made simple power station with one fission reactor, a turbine and an induction matrix. They are all working correctly, however the Reactor PLC and the RTU Gateway have their network light off (disconnected).

Also pocket computers say permission denied when I try to use the alarms app, I think I need to input the auth key, but 

# Computer screenshots

## Reactor PLC
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/2faf0064-7cea-4bc3-85b2-78b6aca1b93f)

### Peripherals
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/504cc36e-506f-48a1-b25d-6ea3ebc5a39b)

The PLC connects to the reactor fine because when I activate in the GUI the PLC automatically turns it off.

## RTU Gateway
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/186ade5d-ac24-46c6-81f6-5d7262b44f11)

### Peripherals
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/cbafeb52-48f8-48a0-95b5-5cfa0970b2c7)

### "SRV" page
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/ce3cc07a-bdd3-4e10-8e84-cb4976194b49)

### "PLC" page
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/46a237ce-d690-4b04-bd7e-9fe156b91bd6)

### "RTU" page
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/7f6910e9-ab65-4b91-b547-4915350ffa60)

### "CRD" page
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/50295f1c-056d-4494-b7a3-0f147c998a6f)

### Peripherals
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/bb52a25d-c929-4467-b03b-2d52d72a8388)

### "PKT" page
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/1f702cab-d663-4a62-8b85-83b9738bcdd5)

### "CRD" page
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/73eab4ea-c310-46f6-b92b-3ee1d67f1bb6)

### "API" page
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/45deb312-3fca-44fe-be79-e0124d518685)

### Peripherals
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/0495c65b-3b45-4160-9710-15437923987e)
 
 
 # Power station setup
 ## Reactor
  
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/6ce5a20d-ec20-44d2-b584-4c8f03f8c4e0)

That network cable goes to the PLC. The waste goes to a bunch of Radioactive Waste Barrels.
 
 ## Turbine
  
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/fb268693-ed11-423b-8df9-72b874c6f3f0)

That network cable goes to the PLC again
 
 ## Induction Matrix

![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/b455c9c3-a807-4d43-abb2-f856d54a292b)

The network cable at the top goes to the RTU Gateway

# Config files
(without real auth key)

[configs.zip](https://github.com/MikaylaFischler/cc-mek-scada/files/13467761/configs.zip)

Some of the configs were made with the GUI configuration programs.
There are no errors in the logs, even with the debug mode enabled.

**Comment by MikaylaFischler** [comment]
Are you using wireless modems or ender modems? Wireless modems have limited range which is reduced by walls and bad weather.

**Reply by samuelh2005** [reply]
All the computers are next to each other with wireless modems, and wired modems to their devices. like this:
![image](https://github.com/MikaylaFischler/cc-mek-scada/assets/41990982/40e0e190-c024-4a77-ac94-aa2473b0b2ec)

**Reply by MikaylaFischler** [reply]
Well the common causes for a completely off network light are
- unmatched authentication keys
- wireless modem range issues
- too short trusted ranges in use
- wrong ports configured

If it's not any of those then I would be stumped without additional information like logs and versions, then I could try to reproduce the problem with the same config files.

The supervisor, PLC, and RTU logs should give some info here if it is none of the above, but given that the network lights are off it seems like it should be one of those four things.

**Reply by samuelh2005** [reply]
My logs (with debug mode enabled) are 
[logs.zip](https://github.com/MikaylaFischler/cc-mek-scada/files/13476035/logs.zip)

My channels are
```lua
-- supervisor comms channel
config.SVR_CHANNEL = 26240
-- PLC comms channel
config.PLC_CHANNEL = 26241
-- RTU/MODBUS comms channel
config.RTU_CHANNEL = 26242
-- coordinator comms channel
config.CRD_CHANNEL = 26243
-- pocket comms channel
config.PKT_CHANNEL = 26244
``` (just the default but with a 2 instead of a 1)

My trusted range is on 0 and my keys are the same. In fact the coordinator and supervisor connect correctly but not the RTU and the PLC.

**Reply by MikaylaFischler** [reply]
I've never seen this happen without user misconfiguration. What I have seen on 1.20.1 is total app crashes on the latest alpha version of ComputerCraft due to switching to Lua 5.2 and actual game crashes with environment detectors. 1.20.1 has been causing headaches.

What version of ComputerCraft are you using?

I can give you some code shortly to add to the comms to trace where things are breaking down.

**Reply by samuelh2005** [reply]
I'm using CC: Tweaked 1.106.1 on 1.201 on Forge 47.2.0

**Reply by MikaylaFischler** [reply]
Well that's the CC version I'm using on CraftOS-PC so it shouldn't be a compatibility issue.

Please add
```lua
log.debug(util.c("plc_nic_tx ", local_channel, " -> ", dest_channel, ": ", textutils.serialize(tx_packet.raw_sendable())))
```
as line 185 in scada-common/network.lua of the reactor-plc computer

And
```lua
log.debug(util.c("plc_handle_pkt ", src_addr, " -> ", l_chan, ": ", textutils.serialize(packet.scada_frame.raw_sendable())))
```
as line 819 in reactor-plc/plc.lua of the reactor-plc computer

And
```lua
log.debug(util.c("svr_handle_pkt ", r_chan, " -> ", l_chan, ": ", textutils.serialize(packet.scada_frame.raw_sendable())))
```
as line 127 in supervisor/supervisor.lua on the supervisor computer

Then start them up and share logs (supervisor & PLC) please after at least 15 seconds of the apps running

**Reply by samuelh2005** [reply]
The new logs are 
[plc-log.txt](https://github.com/MikaylaFischler/cc-mek-scada/files/13477135/plc-log.txt)
[supervisor-log.txt](https://github.com/MikaylaFischler/cc-mek-scada/files/13477158/supervisor-log.txt)

Looks like the supervisor is just ignoring the PLC and RTU

**Reply by MikaylaFischler** [reply]
Please edit `scada-common/network.lua` on the supervisor and uncomment lines 211, 217, and 221 and send logs

**Reply by samuelh2005** [reply]
Looks like the same [supervisor-log.txt](https://github.com/MikaylaFischler/cc-mek-scada/files/13477398/supervisor-log.txt)

**Reply by MikaylaFischler** [reply, marked as answer]
If you are using facility keys and you edited the right file, you should be getting log messages about it verifying the messages if it is receiving any, including from the coordinator.

Looking at that log, I don't see it initiating the cryptography. That means it doesn't have an AUTH_KEY set, so I bet the coordinator is the same way.

I noticed you had named all your configuration files with .settings, I assumed that was just something you edited when zipping them for some reason. Only the Reactor PLC and RTU support the new .settings format, those aren't used elsewhere. The setup guides and wiki instruct that you should still be using `coordinator/config.lua` and `supervisor/config.lua`. If you actually created .settings files or renamed/duplicated them or something like that, then your system is not configured the way you think it is.

**Reply by samuelh2005** [reply]
That was correct. Not sure how the extra .settings files were made. However the server now crashes because the environment detector. Should try to update Advanced Peripherals because I'm on "0.7.32b"
If your interested the crash is [crash-2023-11-27_16.37.56-server.txt](https://github.com/MikaylaFischler/cc-mek-scada/files/13477855/crash-2023-11-27_16.37.56-server.txt).

**Reply by MikaylaFischler** [reply]
Glad you've got it working now! Yeah updating looks like it should work now, I see this in their resolved issues: https://github.com/SirEndii/AdvancedPeripherals/issues/532

**Reply by samuelh2005** [reply]
That fixed it. Thanks!
